src/main/scala/mesosphere/marathon/core/launcher/impl/InstanceOpFactoryImpl.scala

```diff
@@ -354,81 +352,22 @@ class InstanceOpFactoryImpl(
 
     val (reservationLabels, stateOp) = runSpec match {
       case _: AppDefinition =>
-        val hostPorts = resourceMatch.hostPorts.flatten
-        val networkInfo = NetworkInfo(offer.getHostname, hostPorts, ipAddresses = Nil)
-
         // The first taskId does not have an attempt count - this is only the task created to hold the reservation and it
         // will be replaced with a new task once we launch on an existing reservation this way, the reservation will be
         // labeled with a taskId that does not relate to a task existing in Mesos (previously, Marathon reused taskIds so
         // there was always a 1:1 correlation from reservation to taskId)
-        val taskId = Task.Id.forInstanceId(scheduledInstance.instanceId, None)
-        val reservationLabels = TaskLabels.labelsForTask(frameworkId, taskId)
-        val task = Task(
-          taskId = taskId,
-          status = Task.Status(
-            stagedAt = now,
-            condition = Condition.Reserved,
-            networkInfo = networkInfo
-          ),
-          runSpecVersion = runSpec.version
-        )
-
-        val instance = Instance(
-          instanceId = scheduledInstance.instanceId,
-          agentInfo = Some(agentInfo),
-          state = InstanceState(
-            condition = Condition.Reserved,
-            since = now,
-            activeSince = None,
-            healthy = None
-          ),
-          tasksMap = Map(task.taskId -> task),
-          runSpecVersion = runSpec.version,
-          unreachableStrategy = runSpec.unreachableStrategy,
-          reservation = Some(reservation)
-        )
-        val stateOp = InstanceUpdateOperation.Reserve(instance)
+        val reservationLabels = TaskLabels.labelsForTask(frameworkId, Task.Id.forInstanceId(scheduledInstance.instanceId, None))
+        val stateOp = InstanceUpdateOperation.Reserve(Instance.Scheduled(scheduledInstance, reservation, agentInfo))
```

Comment:
This is so much cleaner. Although, I think this would be ok as well

```scala
val reservationLabels = TaskLabels.labelsForTask(frameworkId, Task.Id.forInstanceId(scheduledInstance.instanceId, None))
val updatedInstance = scheduledInstance.copy(reservation = Some(reservation), agentInfo = Some(agentInfo))
val stateOp = InstanceUpdateOperation.Reserve(updatedInstance)
```

Reply:
Push

Reply:
I feel that doing copy actually means that you know too much detail about that particular case class and I don't like that to be spread around code. If you in the end find out that the agent info should be also copied to the tasks in taskmap, you would have to somehow find all the right `.copy` calls and update it... so I tend to like to have `.copy` calls as close to the actual instance as possible :)

Reply:
I agree with @alenkacz here :)

Reply:
Ok. I'm convinced.